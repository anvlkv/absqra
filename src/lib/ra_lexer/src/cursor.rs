use std::str::Chars;
use std::fmt;
pub(crate) const EOF_CHAR: char = '\0';
pub(crate) const EOL_CHAR: char = '\n';

pub(crate) struct Cursor<'a> {
    initial_len: usize,
    chars: Chars<'a>,
    pub position: Position,
    pub level: usize
}

#[derive(Copy, Clone, PartialEq)]
pub struct Position(pub usize, pub usize);

impl fmt::Debug for Position {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "[{}:{}]", self.0, self.1)
    }
}



/// True if `c` is considered a whitespace
pub fn is_whitespace(c: char) -> bool {
    match c {
        // Usual ASCII suspects
        | '\u{0009}' // \t
        | '\u{000B}' // vertical tab
        | '\u{000C}' // form feed
        | '\u{0020}' // space

        // Bidi markers
        | '\u{200E}' // LEFT-TO-RIGHT MARK
        | '\u{200F}' // RIGHT-TO-LEFT MARK
            => true,
        _ => false,
    }
}

/// True if `c` is considered an end of line
pub fn is_end_of_line(c: char) -> bool {
    match c {
        // Usual ASCII suspects
        | '\u{000A}' // \n
        | '\u{000D}' // \r

        // NEXT LINE from latin1
        | '\u{0085}'

        // // Dedicated whitespace characters from Unicode
        | '\u{2028}' // LINE SEPARATOR
        | '\u{2029}' // PARAGRAPH SEPARATOR
            => true,
        _ => false,
    }
}

impl <'a> Cursor<'a> {
    pub(crate)fn new(input: &'a str, position: Position, level: usize) -> Cursor<'a> {
        Cursor {
            initial_len: input.len(),
            chars: input.chars(),
            position,
            level
        }
    }

    /// Returns nth character relative to the current cursor position.
    /// If requested position doesn't exist, `EOF_CHAR` is returned.
    /// However, getting `EOF_CHAR` doesn't always mean actual end of file,
    /// it should be checked with `is_eof` method.
    fn nth_char(&self, n: usize) -> char {
        match self.chars().nth(n).unwrap_or(EOF_CHAR) {
            c if is_end_of_line(c) => EOL_CHAR,
            t => t
        }
    }

    /// Returns a `Chars` iterator over the remaining characters.
    fn chars(&self) -> Chars<'a> {
        self.chars.clone()
    }

    /// Peeks the next symbol from the input stream without consuming it.
    pub(crate) fn first_ahead(&self) -> char {
        self.nth_char(0)
    }

    /// Peeks the second symbol from the input stream without consuming it.
    pub(crate) fn second_ahead(&self) -> char {
        self.nth_char(1)
    }

    /// Checks if there is nothing more to consume.
    pub(crate) fn is_eof(&self) -> bool {
        self.chars.as_str().is_empty()
    }

    /// Moves to the next character.
    pub(crate) fn bump(&mut self) -> Option<char> {
        let character = self.chars.next();
        match character {
            Some(ch) => {
                match ch {
                    c if is_end_of_line(c) => {
                        self.position.0 += 1;
                        self.consume_indent();
                        self.position.1 = self.level;
                        self.bump()
                    }
                    c => {
                        self.position.1 += 1;
                        Some(c)
                    }
                }
            },
            None => None
        }
        
    }

    /// Returns amount of already consumed symbols.
    pub(crate) fn len_consumed(&self) -> usize {
        self.initial_len - self.chars.as_str().len()
    }

    /// Consumes indent level and returns indentation level
    fn consume_indent(&mut self) {
        self.level = self.eat_while(|c| is_whitespace(c));
    }

    /// Eats symbols while predicate returns true or until the end of file is reached.
    /// Returns amount of eaten symbols.
    fn eat_while<F>(&mut self, mut predicate: F) -> usize
    where
        F: FnMut(char) -> bool,
    {
        let mut eaten: usize = 0;
        while predicate(self.first_ahead()) && !self.is_eof() {
            eaten += 1;
            self.bump();
        }

        eaten
    }
}

#[cfg(test)]
mod tests {
    use super::{Cursor, EOF_CHAR, Position};
    #[test]
    fn it_should_create() {
        Cursor::new("abc", Position(1, 0), 0);
    }

    #[test]
    fn it_should_give_first_char() {
        let cur = Cursor::new("abc", Position(1, 0), 0);
        assert_eq!(cur.first_ahead(), 'a');
    }

    #[test]
    fn it_should_give_second_char() {
        let cur = Cursor::new("abc", Position(1, 0), 0);
        assert_eq!(cur.second_ahead(), 'b');
    }

    #[test]
    fn it_should_return_next_char() {
        let mut cur = Cursor::new("a", Position(1, 0), 0);
        assert_eq!(cur.bump().unwrap(), 'a');
    }

    #[test]
    fn it_should_return_none_at_end_of_input() {
        let mut cur = Cursor::new("a", Position(1, 0), 0);
        cur.bump();
        assert_eq!(cur.bump(), None);
    }

    #[test]
    fn it_should_confirm_eof() {
        let mut cur = Cursor::new("a", Position(1, 0), 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.is_eof(), true);
    }

    #[test]
    fn it_should_return_amount_of_consumed_symbols() {
        let mut cur = Cursor::new("abc", Position(1, 0), 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.len_consumed(), 3);
    }

    #[test]
    fn it_should_not_panic_when_encoutering_funny_characters() {
        let mut cur = Cursor::new("ðŸš¬", Position(1, 0), 0);
        assert_eq!(cur.bump().unwrap(), 'ðŸš¬');
    }

    #[test]
    fn it_should_track_position() {
        let mut cur = Cursor::new("abc\nbca", Position(1, 0), 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.position, Position(1, 1));
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.position, Position(1, 2));
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.position, Position(1, 3));
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.position, Position(2, 1));
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.position, Position(2, 2));
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.position, Position(2, 3));
    }

    #[test]
    fn it_should_track_indent_level() {
        let mut cur = Cursor::new("a\n\tb\n\t\tc\nd", Position(1, 0), 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.level, 0);
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.level, 1);
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.level, 2);
        assert_eq!(cur.position, Position(3, 3));
        assert_eq!(cur.bump().unwrap(), 'd');
        assert_eq!(cur.level, 0);
    }
}