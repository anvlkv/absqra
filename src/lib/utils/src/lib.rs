use std::path::Path;
use std::fs::{DirEntry, read_dir};

pub fn files_from_dir_recursively(path: &str) -> Vec<DirEntry> {
    read_dir(Path::new(path)).unwrap().map(|f| f.unwrap()).flat_map(|entry| {
        match entry.metadata().unwrap() {
            meta if meta.is_dir() => files_from_dir_recursively(&entry.path().into_os_string().into_string().unwrap()),
            meta if meta.is_file() => vec![entry],
            _ => panic!("unsupported metadata")
        }
    }).collect()
}