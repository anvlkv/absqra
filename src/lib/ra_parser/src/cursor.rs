use ra_lexer::token::{Token};
use ra_lexer::cursor::Position;
use core::iter::Peekable;

pub(crate) struct Cursor<'token, I> 
    where I: Iterator<Item = Token<'token>> 
{
    pub position: Position,
    pub level: u16,
    tokens: Peekable<I>,
    consumed_len: usize
}

impl <'token, I> Cursor<'token, I> 
where I: Iterator<Item = Token<'token>> 
{
    pub fn new(tokens: I) -> Self
        
    {
        Self {
            position: Position(0, 0),
            level: 0,
            tokens: tokens.peekable(),
            consumed_len: 0
        }
    }

    pub fn first_ahead(&mut self) -> Option<&Token<'token>> {
        self.tokens.peek()
    }

    pub fn bump(&mut self) -> Option<Token<'token>> {
        let tok = self.tokens.next();
        match tok {
            Some(token) => {
                self.position = token.position.1;
                self.level = token.level;
                self.consumed_len += 1;
                Some(token)
            }
            None => None
        }
    }

    pub fn is_eof(& mut self) -> bool {
        self.first_ahead() == None
    }

    // pub fn do_while<F, D>(&mut self, mut test: F, mut cb: D)
    //     where F: FnMut (u16, Position, TokenKind) -> bool,
    //     D: FnMut (Token)
    // {
    //     loop {
    //         let next_token = self.first_ahead();
    //         match next_token {
    //             Some(t) => {
    //                 if test(t.level, t.position.0, t.kind) {
    //                     cb(self.bump().unwrap())
    //                 }
    //                 else {
    //                     break
    //                 }
    //             },
    //             None => break
    //         }
    //     }
    // }

    // pub fn read_within(&mut self, open: TokenKind, close: TokenKind) -> Vec<Token> {
    //     let mut read: Vec<Token> = Vec::new();

    //     loop {
    //         let next_token = self.first_ahead();

    //         match next_token {
    //             Some(tok) => {
    //                 match tok.kind {
    //                     k if k == open => {
    //                         let open_token = self.bump().unwrap();
    //                         let mut nested = self.read_within(open_token.kind, close.clone());
    //                         read.append(&mut nested);
    //                     },
    //                     k if k == close => {
    //                         self.bump();
    //                         break
    //                     },
    //                     _ => read.push(self.bump().unwrap())
    //                 }
    //             },
    //             None => break
    //         }

    //         // println!("{:?}", &read);
    //     }

    //     read
    // }
}