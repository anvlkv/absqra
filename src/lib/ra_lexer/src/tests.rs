use crate::cursor::{Cursor, Position};
use crate::{tokenize, Token, TokenKind};

mod cursor {
    use super::{Cursor, Position};
    #[test]
    fn it_should_create() {
        Cursor::new("abc", Position(1, 0), 0, 0);
    }

    #[test]
    fn it_should_give_first_char() {
        let cur = Cursor::new("abc", Position(1, 0), 0, 0);
        assert_eq!(cur.first_ahead(), 'a');
    }

    #[test]
    fn it_should_give_second_char() {
        let cur = Cursor::new("abc", Position(1, 0), 0, 0);
        assert_eq!(cur.second_ahead(), 'b');
    }

    #[test]
    fn it_should_return_next_char() {
        let mut cur = Cursor::new("a", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
    }

    #[test]
    fn it_should_return_none_at_end_of_input() {
        let mut cur = Cursor::new("a", Position(1, 0), 0, 0);
        cur.bump();
        assert_eq!(cur.bump(), None);
    }

    #[test]
    fn it_should_confirm_eof() {
        let mut cur = Cursor::new("a", Position(1, 0), 0, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.is_eof(), true);
    }

    #[test]
    fn it_should_return_amount_of_consumed_symbols() {
        let mut cur = Cursor::new("abc", Position(1, 0), 0, 0);
        cur.bump();
        cur.bump();
        cur.bump();
        assert_eq!(cur.len_consumed(), 3);
    }

    #[test]
    fn it_should_not_panic_when_encountering_funny_characters() {
        let mut cur = Cursor::new("ðŸš¬", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'ðŸš¬');
    }

    #[test]
    fn it_should_track_position() {
        let mut cur = Cursor::new("abc\nSOME", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.position, Position(1, 1));
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.position, Position(1, 2));
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.position, Position(1, 3));
        assert_eq!(cur.bump().unwrap(), 'S');
        assert_eq!(cur.position, Position(2, 1));
        assert_eq!(cur.bump().unwrap(), 'O');
        assert_eq!(cur.position, Position(2, 2));
        assert_eq!(cur.bump().unwrap(), 'M');
        assert_eq!(cur.position, Position(2, 3));
    }

    #[test]
    fn it_should_track_indent_level() {
        let mut cur = Cursor::new("a\n\tb\n\t\tc\nd", Position(1, 0), 0, 0);
        assert_eq!(cur.bump().unwrap(), 'a');
        assert_eq!(cur.level, 0);
        assert_eq!(cur.bump().unwrap(), 'b');
        assert_eq!(cur.level, 1);
        assert_eq!(cur.bump().unwrap(), 'c');
        assert_eq!(cur.level, 2);
        assert_eq!(cur.position, Position(3, 3));
        assert_eq!(cur.bump().unwrap(), 'd');
        assert_eq!(cur.level, 0);
    }
}

mod lib {
    use super::{tokenize, Position, Token, TokenKind};
    

    #[test]
    fn it_should_create_iterator_of_tokens() {
        let mut stream = tokenize("abc");
        stream.next();
    }

    #[test]
    fn it_should_parse_multiple_lines() {
        let mut stream = tokenize("!?&/\n\tabc");
        stream.next();
        stream.next();
        stream.next();
        stream.next();
        assert_eq!(
            stream.next().unwrap(),
            Token {
                kind: Some(TokenKind::Identifier("abc")),
                content: "abc",
                position: (Position(2, 1), Position(2, 4)),
                len: 3,
                level: 1,
            }
        )
    }

    #[test]
    fn it_should_parse_identifiers() {
        let mut stream = tokenize("abc");
        assert_eq!(
            stream.next().unwrap(),
            Token {
                kind: Some(TokenKind::Identifier("abc")),
                content: "abc",
                position: (Position(1, 0), Position(1, 3)),
                len: 3,
                level: 0
            }
        );
    }

    use ra_common::insta::assert_debug_snapshot;
    use utils::for_each_ra_example_file;
    use std::fs::File;
    use std::io::Read;

    #[test]
    fn it_should_match_snapshots() {
        for_each_ra_example_file(|example| {
                let mut file = File::open(example.path()).unwrap();
    
                let mut contents = String::new();
                file.read_to_string(&mut contents).unwrap();

                println!("{}", example.file_name().to_str().unwrap());

                let tokens: Vec<Token> = tokenize(&contents).collect();
                assert_debug_snapshot!(example.path().to_str().unwrap(), tokens)
        });
    }

    mod strings {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_string_literals_with_double_quotes() {
            let mut stream = tokenize("\"some\"");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::StringLiteral("some")),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    ..Default::default()
                }
            )
        }

        #[test]
        fn it_should_parse_string_literals_with_single_quotes() {
            let mut stream = tokenize("\'some\'");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::StringLiteral("some")),
                    content: "some",
                    len: 6,
                    position: (Position(1, 0), Position(1, 6)),
                    ..Default::default()
                }
            )
        }
    }
    mod comments {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_single_line_comments() {
            let mut stream = tokenize("//abc");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 0
                }
            );
        }
        #[test]
        fn it_should_parse_multi_line_comments() {
            let mut stream = tokenize("/*abc\nSOME*/");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "abc\nSOME",
                    position: (Position(1, 0), Position(2, 6)),
                    len: 12,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_multi_line_comments_with_fringe() {
            let mut stream = tokenize("/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "\n* abc\n* SOME\n",
                    position: (Position(1, 0), Position(4, 2)),
                    len: 18,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_comments_followed_by_one_another() {
            let mut stream = tokenize("//some\n\n/*\n* abc\n* SOME\n*/");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "some",
                    position: (Position(1, 0), Position(1, 6)),
                    len: 6,
                    level: 0
                }
            );

            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Comment),
                    content: "\n* abc\n* SOME\n",
                    position: (Position(3, 0), Position(6, 2)),
                    len: 18,
                    level: 0
                }
            );
        }
    }
    
    mod numbers {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_numbers() {
            let mut stream = tokenize("123");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Int(123)),
                    content: "123",
                    position: (Position(1, 0), Position(1, 3)),
                    len: 3,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_separator() {
            let mut stream = tokenize("123,321");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123.321)),
                    content: "123,321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_separator() {
            let mut stream = tokenize("123.321");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123.321)),
                    content: "123.321",
                    position: (Position(1, 0), Position(1, 7)),
                    len: 7,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_thousands_separator() {
            let mut stream = tokenize("123.321,456");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123321.456)),
                    content: "123.321,456",
                    position: (Position(1, 0), Position(1, 11)),
                    len: 11,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_decimal_and_multiple_thousands_separator() {
            let mut stream = tokenize("123.321.123,456");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123321123.456)),
                    content: "123.321.123,456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_parse_numbers_with_another_decimal_and_multiple_thousands_separators() {
            let mut stream = tokenize("123,321,123.456");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(123321123.456)),
                    content: "123,321,123.456",
                    position: (Position(1, 0), Position(1, 15)),
                    len: 15,
                    level: 0
                }
            );
        }

        #[test]
        #[should_panic]
        fn it_should_panic_when_encountering_multiple_decimal_separators() {
            let mut stream = tokenize("123.321.123,456,654");
            stream.next();
        }

        #[test]
        fn it_should_parse_negative_integer() {
            let mut stream = tokenize("-123");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Int(-123)),
                    content: "-123",
                    position: (Position(1, 0), Position(1, 4)),
                    len: 4,
                    level: 0
                }
            )
        }

        #[test]
        fn it_should_parse_negative_float() {
            let mut stream = tokenize("-123.312");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Float(-123.312)),
                    content: "-123.312",
                    position: (Position(1, 0), Position(1, 8)),
                    len: 8,
                    level: 0
                }
            )
        }
    }

    mod content_block {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_content_blocks() {
            let mut stream = tokenize("`abc`");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::ContentBlock),
                    content: "abc",
                    position: (Position(1, 0), Position(1, 5)),
                    len: 5,
                    level: 0
                }
            );
        }

        #[test]
        fn it_should_keep_inner_indents_when_parsing_content() {
            let mut stream = tokenize("`\n\tabc\n\t\tabc\n`");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::ContentBlock),
                    content: "\n\tabc\n\t\tabc\n",
                    position: (Position(1, 0), Position(4, 1)),
                    len: 14,
                    level: 0
                }
            );
        }

        #[test]
        #[should_panic]
        fn it_should_panic_when_content_block_is_not_closed_properly() {
            let mut stream = tokenize("`\n\tabc`");
            stream.next();
        }
    }

    mod single_char_tokens {
        use super::{tokenize, Position, Token, TokenKind};
        #[test]
        fn it_should_parse_tokens() {
            let mut stream = tokenize("!?&/#");
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Exclamation),
                    position: (Position(1, 0), Position(1, 1)),
                    content: "!",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Question),
                    position: (Position(1, 1), Position(1, 2)),
                    content: "?",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Ampersand),
                    position: (Position(1, 2), Position(1, 3)),
                    content: "&",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::Slash),
                    position: (Position(1, 3), Position(1, 4)),
                    content: "/",
                    level: 0,
                    ..Default::default()
                }
            );
            assert_eq!(
                stream.next().unwrap(),
                Token {
                    kind: Some(TokenKind::HashPound),
                    position: (Position(1, 4), Position(1, 5)),
                    content: "#",
                    level: 0,
                    ..Default::default()
                }
            );
        }
    }


}
