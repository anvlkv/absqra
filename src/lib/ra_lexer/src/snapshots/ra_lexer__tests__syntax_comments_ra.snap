---
source: src/lib/ra_lexer/src/tests.rs
expression: tokens
---
[
  {
    "kind": {
      "Comment": [
        " simply a one line comment",
        false
      ]
    },
    "position": [
      [
        1,
        1
      ],
      [
        1,
        29
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " fancy one line comment",
        true
      ]
    },
    "position": [
      [
        3,
        1
      ],
      [
        3,
        27
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        "\n fancy multi line comment\n \n",
        true
      ]
    },
    "position": [
      [
        5,
        1
      ],
      [
        8,
        4
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        "\nsimply \na multi line\ncomment\n",
        false
      ]
    },
    "position": [
      [
        10,
        1
      ],
      [
        14,
        3
      ]
    ],
    "level": 1
  },
  {
    "kind": {
      "Comment": [
        " –Ω–∞ —ç—Ç–æ–º –≤—Å—ë, —Ä–µ–±—è—Ç–∞! üòá",
        false
      ]
    },
    "position": [
      [
        16,
        1
      ],
      [
        16,
        26
      ]
    ],
    "level": 1
  }
]
